### PR TITLE
prospective-parachains rework: take II

### PR DESCRIPTION
**Pull request by alindima**
Resolves https://github.com/paritytech/polkadot-sdk/issues/4800

# Problem
In https://github.com/paritytech/polkadot-sdk/pull/4035, we removed support for parachain forks and cycles and added support for backing unconnected candidates (candidates for which we don't yet know the full path to the latest included block), which is useful for elastic scaling (parachains using multiple cores).

Removing support for backing forks turned out to be a bad idea, as there are legitimate cases for a parachain to fork (if they have other consensus mechanism for example, like BABE or PoW). This leads to validators getting lower backing rewards (depending on whether they back the winning fork or not) and a higher pressure on only the half of the backing group (during availability-distribution for example). Since we don't yet have approval voting rewards, backing rewards are a pretty big deal (which may change in the future).

# Description

A backing group is now allowed to back forks. Once a candidate becomes backed (has the minimum backing votes), we don't accept new forks unless they adhere to the new fork selection rule (have a lower candidate hash).
This helps with keeping the implementation simpler, since forks will only be taken into account for candidates which are not backed yet (only seconded).
Having this fork selection rule also helps with reducing the work backing validators need to do, since they have a shared way of picking the winning fork. Once they see a candidate backed, they can all decide to back a fork and not accept new ones.
But they still accept new ones during the seconding phase (until the backing quorum is reached).

Therefore, a block author which is not part of the backing group will likely not even see the forks (only the winning one).

Just as before, a parachain producing forks will still not be able to leverage elastic scaling but will still work with a single core. Also, cycles are still not accepted.

## Some implementation details

`CandidateStorage` is no longer a subsystem-wide construct. It was previously holding candidates from all relay chain forks and complicated the code. Each fragment chain now holds their candidate chain and their potential candidates. This should not increase the storage consumption since the heavy candidate data is already wrapped in an Arc and shared. It however allows for great simplifications and increase readability.

`FragmentChain`s are now only creating a chain with backed candidates and the fork selection rule. As said before, `FragmentChain`s are now also responsible for maintaining their own potential candidate storage.

Since we no longer have the subsytem-wide `CandidateStorage`, when getting a new leaf update, we use the storage of our latest ancestor, which may contain candidates seconded/backed that are still in scope.

When a candidate is backed, the fragment chains which hold it are recreated (due to the fork selection rule, it could trigger a "reorg" of the fragment chain).

I generally tried to simplify the subsystem and not introduce unneccessary optimisations that would otherwise complicate the code and not gain us much (fragment chains wouldn't realistically ever hold many candidates)

TODO:
- [x] update metrics
- [x] update docs and comments
- [x] fix and add unit tests
- [x] tested with fork-producing parachain
- [x] tested with cycle-producing parachain
- [x] versi test
- [x] prdoc

### CODE REVIEW

**Comment by alindima** [comment]
This is ready for an initial review. CI zombienet tests pass and all of my local zombienet tests work well.
I'm working on adding unit tests and after that I'll test on versi.

In the meantime, it would be good to get some in-depth reviews. I'd appreciate very much reviews that focus on the protocol changes and the design rather than nitpicks about function names and structure (these can come at a later stage).

CC: @sandreim @eskimor @alexggh @tdimitrov

**Comment by tdimitrov** [comment]
Great work, Alin! I did one pass focusing mainly on the logic and the PR looks good. I'll  do another full pass when you are ready.

**Review by sandreim — COMMENTED** [review]
Took a first pass on the changes. Looks good in general, I could not find any logic errors wrt the introduced changes in behaviour. However I do appreciate the changes that go deeper into the original prospective parachain subsystem architecture. Now things are much better to understand.

I

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:879]: what are you referring to?

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:366]: There are multiple cases where this function could fail. I think it's a good practice to just have the user just look at the possible Error variants, rather than mentioning them in doc comments (which can easily become outdated)

**Comment by alindima** [comment]
I took this PR out of draft mode. The only thing left to do is to test it for a while on versi.
So it's ready for the second round of reviews/approves. @eskimor @sandreim @tdimitrov

File: polkadot/node/core/prospective-parachains/src/lib.rs
Comment (alindima) [inline polkadot/node/core/prospective-parachains/src/lib.rs:368]: There's a bug here. We should keep deactivated leaves until they go out of scope (like backing_implicit_view) does, so that a subsequent fork may get the same candidates from the parent as the first fork did.

Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/lib.rs]: fixed

**Comment by alindima** [comment]
Versi testing was successful ✅

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (eskimor) [inline polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:1239]: While elegant, also quite inefficient. We do full rebuilding and pruning of store (breath first search) on each and every backed candidate. With many cores (NOMT, other PoV bound usecases) this adds up and I think it is also not necessary.

Instead of doing `Self::populate()`, we should do:

```rust
self.revert_chain_to(head_data: Hash); // newly introduced method traversing the chain backwards, dropping all elements (moving back to storage) until candidate with given head data is found.
self.populate_chain();
```

This way:

1. We don't do costly pruning on each and every backing.
2. Most likely case that the chain is built from start to finish, `revert_chain_to` will do nothing and `populate_chain` has very little work.
3. Unlikely case that we build in reverse order, we are still kind of efficient as there is at least no chain to traverse and we only got once through the storage to find a potential child.
4. In the more realistic case that candidates simply come slightly out of order, we get a mixture of (2) and (3), which should also be quite efficient.

File: polkadot/node/core/prospective-parachains/src/lib.rs
Comment (eskimor) [inline polkadot/node/core/prospective-parachains/src/lib.rs:354]: This function does not have access to scope and does not check it at all. The only thing it checks whether the candidate is present already.

File: polkadot/node/network/collator-protocol/src/validator_side/collation.rs
Comment (eskimor) [inline polkadot/node/network/collator-protocol/src/validator_side/collation.rs:153]: The prioritization only happens when fully backed. So this seems unrelated?

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:115]: Overseer requires the main subsystem future to be `Send`. The prospective-parachains is not `Send` if we use an `Rc` here, since this `Rc` is being held across `await` points in the code.

File: polkadot/node/core/prospective-parachains/src/lib.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/lib.rs:354]: correct, the comment is misleading

File: polkadot/node/network/collator-protocol/src/validator_side/collation.rs
Comment (alindima) [reply polkadot/node/network/collator-protocol/src/validator_side/collation.rs:153]: Even if the prioritisation happens when fully backed, it may be misused to prioritise a candidate. I'm not sure we check the candidate's hash against the provided candidate hash somewhere else (during validation).

We do this check here and in statement distribution and punish the reputation if misused.

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:790]: Due to the way the provisioner is coded, this can happen if a candidate times out availability.

The provisioner does not do any ordering on the ancestors (which would increase complexity, as it doesn't have the right data yet and the candidates are not occupying the cores in order). So if a candidates times out at this block, it will not add it to the ancestors set (but it will add the descendants of the timed out candidate).

That's why we stop when we find a hole

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs]: good idea. done

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:694]: I replaced one use of `as_candidate_storage` (when marking a candidate as backed) with your suggestion.
However, the other place where we call this can't consume it (during active leaf update), because subsequent forks need to have access to the parent's candidates. I'll see how I can make this prettier (maybe by using your advance_scope suggestion)

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs]: An initial prototype of mine did something similar. However, I decided that the added complexity was not worth it and I tried to keep it simple. Fair point, I'll reconsider this

File: polkadot/node/core/prospective-parachains/src/lib.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/lib.rs:281]: I refactored this code a bit taking into account your comments. I didn't take all of this code out of active_leaves_update, but I think it's best like this. Let me know if it looks good

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs]: Done, I optimised this. Have a look!

File: polkadot/node/network/collator-protocol/src/validator_side/collation.rs
Comment (eskimor) [reply polkadot/node/network/collator-protocol/src/validator_side/collation.rs:153]: I guess my point is: Advertising a hash and delivering something else is malicious. Trying to manipulate the fork choice rule would be the least of my worries.

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (eskimor) [inline polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:728]: This function does not seem to actually advance the scope (move to the child relay chain block).

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (eskimor) [inline polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:1063]: Some docs telling what that starting point actually is would be nice.

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (eskimor) [inline polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:1338]: Hmm. Maybe we should have a `CandidateEntry::from(FragmentNode)` instance?

**Review by eskimor — COMMENTED** [review]
Will have a closer look tomorrow, from a quick pass it seems that `advance_scope` does not actually advance scope, but instead the advancing logic still remains in `handle_active_leaves_update`. You mentioned that and there may be good reasons, but the very least the function name seems off then.

File: polkadot/node/core/prospective-parachains/src/lib.rs
Comment (eskimor) [reply polkadot/node/core/prospective-parachains/src/lib.rs:281]: Let's discuss tomorrow.

File: polkadot/node/network/collator-protocol/src/validator_side/collation.rs
Comment (alindima) [reply polkadot/node/network/collator-protocol/src/validator_side/collation.rs:153]: Ok I get what you mean now. Fair point

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:1338]: I wouldn't add that, as there is no obvious way of telling what `CandidateState` (backed or seconded) to use without additional context

Comment (eskimor) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:1338]: Why not: I believe the invariant is that everything that is on a chain is that it was backed?

FragmentNode -> CandidateEntry should be good.


I am assuming here that a `FragmentNode` is an element of `BackedChain` ... although I am mostly inferring this from the code, this invariant is actually not documented.

File: polkadot/node/core/prospective-parachains/src/lib.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/lib.rs]: Refactored this as discussed. Have a look!

File: polkadot/node/network/collator-protocol/src/validator_side/collation.rs
Comment (alindima) [reply polkadot/node/network/collator-protocol/src/validator_side/collation.rs]: removed this comment

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs]: done

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (eskimor) [inline polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:834]: Best to leave that decision to the caller. If we need a CandidateEntry and not a &CandidateEntry, we should just state it so and leave it to the caller whether any cloning is required or not.

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (eskimor) [inline polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:688]: Reminder.

**Review by eskimor — APPROVED** [review]
Nice! Approving modulo the todo.

**Review by sandreim — APPROVED** [review]
Excellent work!

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs:834]: the idea was to not clone unless we have to. There's no need to clone if the `self.can_add_candidate_as_potential(candidate)?;` returns an Error

File: polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs
Comment (alindima) [reply polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs]: true. I updated the docs

**Pull request merged**
